tests/hazmat/primitives/test_pkcs12.py

```diff
@@ -137,3 +139,113 @@ def test_buffer_protocol(self, backend):
         assert parsed_key is not None
         assert parsed_cert is not None
         assert parsed_more_certs == []
+
+
+def _load_cert(backend, path):
+    return load_vectors_from_file(
+        path,
+        lambda pemfile: x509.load_pem_x509_certificate(
+            pemfile.read(), backend
+        ), mode='rb'
+    )
+
+
+def _load_ca(backend):
+    cert = _load_cert(backend, os.path.join('x509', 'custom', 'ca', 'ca.pem'))
+    key = load_vectors_from_file(
+        os.path.join('x509', 'custom', 'ca', 'ca_key.pem'),
+        lambda pemfile: load_pem_private_key(
+            pemfile.read(), None, backend
+        ), mode='rb'
+    )
+    return cert, key
+
+
+def _load_alt_cert(backend):
+    return _load_cert(
+        backend, os.path.join('x509', 'custom', 'dsa_selfsigned_ca.pem')
+    )
+
+
+class TestPKCS12Creation(object):
+    @pytest.mark.parametrize('name', [None, b'name'])
+    @pytest.mark.parametrize(('key_encryption', 'password'), [
+        (serialization.BestAvailableEncryption(b'password'), b'password'),
+        (serialization.NoEncryption(), None)
+    ])
+    def test_generate(self, backend, name, key_encryption, password):
+        cert, key = _load_ca(backend)
+        p12 = serialize_key_and_certificates(
+            name, key, cert, None, key_encryption)
+
+        parsed_key, parsed_cert, parsed_more_certs = \
+            load_key_and_certificates(p12, password, backend)
+        assert parsed_cert == cert
+        assert parsed_key.private_numbers() == key.private_numbers()
+        assert parsed_more_certs == []
+
+    def test_generate_with_ca(self, backend):
+        cert, key = _load_ca(backend)
+        cert2 = _load_alt_cert(backend)
+        encryption = serialization.NoEncryption()
+        p12 = serialize_key_and_certificates(
+            None, key, cert, [cert2], encryption)
+
+        parsed_key, parsed_cert, parsed_more_certs = \
+            load_key_and_certificates(p12, None, backend)
+        assert parsed_cert == cert
+        assert parsed_key.private_numbers() == key.private_numbers()
+        assert parsed_more_certs == [cert2]
+
+        with pytest.raises(TypeError) as exc:
+            serialize_key_and_certificates(None, key, cert, cert2, encryption)
+
+        with pytest.raises(TypeError) as exc:
+            serialize_key_and_certificates(None, key, cert, [key], encryption)
+        assert str(exc.value) == 'cert in cas must be a certificate'
```

Comment:
We don't usually do these assertions, do we?

Reply:
What do you mean exactly? The check for assertion message?

Reply:
We're inconsistent about it, but I'm fine with leaving them here so they're still in the latest commit.